src/varname.jl

```diff
@@ -396,80 +405,109 @@ Currently _not_ supported are:
 subsumes_index(::Tuple{}, ::Tuple{}) = true  # x subsumes x
 subsumes_index(::Tuple{}, ::Tuple) = true    # x subsumes x[1]
 subsumes_index(::Tuple, ::Tuple{}) = false   # x[1] does not subsume x
-function subsumes_index(t::Tuple, u::Tuple)  # does x[i]... subsume x[j]...?
-    return _issubindex(first(t), first(u)) && subsumes_index(Base.tail(t), Base.tail(u))
+function subsumes_index(t1::Tuple, t2::Tuple)  # does x[i]... subsume x[j]...?
+    first_subsumed = all(zip(first(t1), first(t2))) do (i, j)
+        if j isa Colon
+            error("Colons cannot be subsumed")
+        elseif i isa Colon
+            return true
+        else
+            return issubset(j, i)
```

Comment:
This will fail for keys that aren't integers, e.g.

``` julia
julia> vn1 = @varname(x[:a]); vn2 = @varname(x[:a][1]);

julia> vn1 ≍ vn2
ERROR: MethodError: no method matching length(::Symbol)
Closest candidates are:
  length(::Union{Base.KeySet, Base.ValueIterator}) at /opt/julia-1.7.0/share/julia/base/abstractdict.jl:58
  length(::Union{LinearAlgebra.Adjoint{T, S}, LinearAlgebra.Transpose{T, S}} where {T, S}) at /opt/julia-1.7.0/share/julia/stdlib/v1.7/LinearAlgebra/src/adjtrans.jl:171
  length(::Union{Tables.AbstractColumns, Tables.AbstractRow}) at ~/.julia/packages/Tables/M26tI/src/Tables.jl:175
  ...
Stacktrace:
  [1] issubset(l::Symbol, r::Symbol)
    @ Base /opt/julia-1.7.0/share/julia/base/abstractset.jl:281
  [2] (::AbstractPPL.var"#15#16")(::Tuple{Symbol, Symbol})
    @ AbstractPPL ~/Projects/public/AbstractPPL.jl/src/varname.jl:415
  [3] _all(f::AbstractPPL.var"#15#16", itr::Base.Iterators.Zip{Tuple{Tuple{Symbol}, Tuple{Symbol}}}, #unused#::Colon)
    @ Base /opt/julia-1.7.0/share/julia/base/reduce.jl:1157
  [4] all(f::Function, itr::Base.Iterators.Zip{Tuple{Tuple{Symbol}, Tuple{Symbol}}})
    @ Base /opt/julia-1.7.0/share/julia/base/reduce.jl:1152
  [5] subsumes_index(t1::Tuple{Tuple{Symbol}, Tuple{Int64}}, t2::Tuple{Tuple{Symbol}})
    @ AbstractPPL ~/Projects/public/AbstractPPL.jl/src/varname.jl:409
  [6] subsumes_index(t::Setfield.ComposedLens{Setfield.IndexLens{Tuple{Symbol}}, Setfield.IndexLens{Tuple{Int64}}}, u::Setfield.IndexLens{Tuple{Symbol}})
    @ AbstractPPL ~/Projects/public/AbstractPPL.jl/src/varname.jl:361
  [7] subsumes(t::Setfield.ComposedLens{Setfield.IndexLens{Tuple{Symbol}}, Setfield.IndexLens{Tuple{Int64}}}, u::Setfield.IndexLens{Tuple{Symbol}})
    @ AbstractPPL ~/Projects/public/AbstractPPL.jl/src/varname.jl:304
  [8] subsumes(u::VarName{:x, Setfield.ComposedLens{Setfield.IndexLens{Tuple{Symbol}}, Setfield.IndexLens{Tuple{Int64}}}}, v::VarName{:x, Setfield.IndexLens{Tuple{Symbol}}})
    @ AbstractPPL ~/Projects/public/AbstractPPL.jl/src/varname.jl:269
  [9] subsumedby(t::VarName{:x, Setfield.IndexLens{Tuple{Symbol}}}, u::VarName{:x, Setfield.ComposedLens{Setfield.IndexLens{Tuple{Symbol}}, Setfield.IndexLens{Tuple{Int64}}}})
    @ AbstractPPL ~/Projects/public/AbstractPPL.jl/src/varname.jl:307
 [10] ⋢(t::VarName{:x, Setfield.IndexLens{Tuple{Symbol}}}, u::VarName{:x, Setfield.ComposedLens{Setfield.IndexLens{Tuple{Symbol}}, Setfield.IndexLens{Tuple{Int64}}}})
    @ AbstractPPL ~/Projects/public/AbstractPPL.jl/src/varname.jl:311
 [11] ≍(t::VarName{:x, Setfield.IndexLens{Tuple{Symbol}}}, u::VarName{:x, Setfield.ComposedLens{Setfield.IndexLens{Tuple{Symbol}}, Setfield.IndexLens{Tuple{Int64}}}})
    @ AbstractPPL ~/Projects/public/AbstractPPL.jl/src/varname.jl:312
 [12] top-level scope
    @ REPL[14]:1
```

More specifically

``` julia
julia> issubset(:a, :a)
ERROR: MethodError: no method matching length(::Symbol)
Closest candidates are:
  length(::Union{Base.KeySet, Base.ValueIterator}) at /opt/julia-1.7.0/share/julia/base/abstractdict.jl:58
  length(::Union{LinearAlgebra.Adjoint{T, S}, LinearAlgebra.Transpose{T, S}} where {T, S}) at /opt/julia-1.7.0/share/julia/stdlib/v1.7/LinearAlgebra/src/adjtrans.jl:171
  length(::Union{Tables.AbstractColumns, Tables.AbstractRow}) at ~/.julia/packages/Tables/M26tI/src/Tables.jl:175
  ...
Stacktrace:
 [1] issubset(l::Symbol, r::Symbol)
   @ Base /opt/julia-1.7.0/share/julia/base/abstractset.jl:281
 [2] top-level scope
   @ REPL[15]:1
```

Though this should also be broken in the implementation currently present on `#master`.

Reply:
One solution is to split the if-statement into different dispatches for some function, and then add a default `f(x, y) = x == y` or something.

Reply:
I know, this is somethink I monkey-patched again to make tests work. Originally I wanted `checkindex`, but that unfortunately throws an error in cases it is not meant for and is even worse. Dispatch sounds good; maybe we can come up with a reasonably extensible interface here.